ユーザー管理機能

kumazakiz opened the pull request:
#WHAT
-ユーザー管理機能の実装

#WHY
-新規登録機能を実装するため
-ログイン・ログアウト機能を実装するため

-追記
　-手順を間違えて、デプロイを先にしてしまいました。

-新規登録が正常にできる
https://gyazo.com/9e21551adda2f069670796bcd9b84acc
-新規登録に失敗した場合はエラーが表示される
https://gyazo.com/1aa3227946a75753fefe525c09a36809
-ログイン
https://gyazo.com/8947ce9e4ff924279d220f1a6859a157
-ログアウト
https://gyazo.com/98cb9e401e9cad7a145c86c4b51d2279
-テストコードの動作確認
https://gyazo.com/1139484603c51770a5bc14f03ed2c590

kumazakiz committed:
Deviseの導入

kumazakiz committed:
userモデルの作成

kumazakiz committed:
usersテーブルの作成

kumazakiz committed:
usersコントローラーの作成

kumazakiz committed:
imageファイルの配置

kumazakiz committed:
deviseのビューファイルの作成

kumazakiz committed:
新規投稿機能の実装(1)

kumazakiz committed:
reset.cssの配置

kumazakiz committed:
stylesheetsの配置

kumazakiz committed:
新規登録機能の実装(2)

kumazakiz committed:
ログイン・ログアウトの実装

kumazakiz committed:
userモデルの単体テストコードの実装

kumazakiz committed:
バリデーションの追記

kumazakiz committed:
userモデルの単体テストコードの実装(2)

[userモデルの単体テストコードの実装]の続きの作業分です。

kumazakiz committed:
ファイルの変更履歴をHerokuへ反映させる

内容
rubocopコマンドでのチェック内容の自動修正分です。

i-date commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,10 @@
+  private
+
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:nickname, :email, :encrypted_password, :family_name, :first_name, :family_name_kana, :first_name_kana, :birthday])
```
emailとencrypted_passwordについてはdeviseにてデフォルトで扱われるため、8行目に記述する必要はありません。
そのため、8行目の`:email`と`:encrypted_password`は削除しましょう。

i-date commented on app/models/user.rb:
```diff
@@ -0,0 +1,17 @@
+  PASSWORD_REGEX = /\A(?=.*?[a-z])(?=.*?\d)[a-z\d]+\z/i.freeze
+  validates_format_of :password, with: PASSWORD_REGEX
+  validates :nickname,         presence: true
+  with_options presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]+\z/, message: '全角文字を使用してください' } do
+    validates :family_name
+    validates :first_name
+    validates :family_name_kana
+    validates :first_name_kana
```
family_name_kanaとfirst_name_kanaも、10行目の「全角（漢字・ひらがな・カタカナ）」を表現する正規表現でまとめられていますが、family_name_kanaとfirst_name_kanaは「全角（カタカナ）」になりますので、別の正規表現のバリデーションでまとめましょう。

i-date commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,99 @@
+        @user.email = '@'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Email is invalid')
+      end
```
ユーザー管理機能の実装条件が満たされているか確認するため、emailについて下記の場合のテストも追加しましょう。
- emailが既に登録済みである場合は登録できない（メールアドレスが一意性であること）

i-date reviewed (commented):
レビュー依頼ありがとうございます！
コメントいたしましたので、ご確認をお願いいたします。

i-date commented:
また恐れ入りますが、実装いただいた機能動作や画面を確認させていただきたいので、上記修正後に、プルリクに下記の項目を確認できるキャプチャ動画（のURL）を載せていただき、再度プルリクエストをご提出いただけないでしょうか？
- 新規登録が正常にできる
- 新規登録に失敗した場合はエラーが表示される
- ログイン
- ログアウト

kumazakiz committed:
コードレビューで指摘された部分の修正分

①application_controller.rbの修正
②models/user.rbのバリデーションの修正
③user_spec.rbにテストコードの追加

masayaomoto commented on app/models/user.rb:
```diff
@@ -0,0 +1,19 @@
+    validates :family_name_kana
+    validates :first_name_kana
+  end
+  validates :birthday, presence: true
```
9行目```validates :nickname,         presence: true```
18行目```validates :birthday, presence: true```
これらは```with_options```でまとめることができそうです！
ご確認をお願いいたします！

masayaomoto commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,106 @@
+        @user.password = 'aaaaaa'
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password confirmation doesn't match Password", 'Password is invalid')
+      end
```
テストの正確性を向上するため、「passwordが半角英数字混合入力でないと登録できない」というテストは下記の2つのテストに分けましょう！
```
passwordが半角数字のみの場合は登録できない
passwordが半角英字のみの場合は登録できない
```

masayaomoto reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

kumazakiz committed:
コードレビューで指摘された部分の修正(2)

①models/user.rbのバリデーション[nickname]、[birthday]を一つにまとめる。

②models/user_spec.rbにpasswordのテストコードを二つ追加

kokikurahashi commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,112 @@
+        @user.valid?
+        expect(@user.errors.full_messages).to include("Password can't be blank")
+      end
+      it 'passwordが5文字以下だと登録できない' do
+        @user.password = '12345'
+        @user.password_confirmation = '12345'
+        @user.valid?
+        expect(@user.errors.full_messages).to include('Password is too short (minimum is 6 characters)')
+      end
+      it 'passwordが半角数字のみの場合は登録できない' do
```
passwordの字数制限のテストでは「英数字混合だが、字数が5文字以下」の場合で行った方が好ましいです！

kokikurahashi reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
一点追加でコメントいたしましたので、ご確認お願いします！

kumazakiz committed:
コードレビューで指摘された部分の修正

①models/user_spec.rbに[passwordは英数字混合だが、5文字以下だと登録できない]の項目を追加

i-date reviewed (approved):
修正ありがとうございます！
修正点を拝見したところ問題ないことを確認できましたので、LGTMとさせていただきます。
以降の実装も引き続きこの調子で進めていきましょう。応援しております！
